pkg/operator/apiserver/controller/apiservice/apiservice_controller_test.go

```diff
@@ -412,6 +437,89 @@ func TestDisabledAPIService(t *testing.T) {
 
 }
 
+func TestPreconditionsForEnabledAPIServices(t *testing.T) {
+	var (
+		bootstrapComplete = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Name: "bootstrap", Namespace: "kube-system"},
+			Data:       map[string]string{"status": "complete"},
+		}
+
+		bootstrapNoStatus = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Name: "bootstrap", Namespace: "kube-system"},
+			Data:       map[string]string{"noStatus": "complete"},
+		}
+	)
+
+	scenarios := []struct {
+		name                       string
+		existingBootstrapConfigMap *corev1.ConfigMap
+		existingAPIServices        []*apiregistrationv1.APIService
+		existingEndpoints          []*corev1.Endpoints
+
+		expectedStatus bool
+	}{
+		{
+			name: "EmptyAPIServices, NoBootstrapConfigMap",
+		},
+		{
+			name:                       "EmptyAPIServices, IncompleteBootstrapConfigMap",
+			existingBootstrapConfigMap: bootstrapNoStatus,
+		},
+		{
+			name:                       "EmptyAPIServices, CompleteBootstrapConfigMap",
+			existingBootstrapConfigMap: bootstrapComplete,
+			expectedStatus:             true,
+		},
+		{
+			name: "UnreadyAPIServices, CompleteBootstrapConfigMap",
+			existingAPIServices: []*apiregistrationv1.APIService{
+				newAPIService("build.openshift.io", "v1"),
+			},
+			existingEndpoints: []*corev1.Endpoints{
+				func() *corev1.Endpoints {
+					s := newAPIService("build.openshift.io", "v1")
+					e := &corev1.Endpoints{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      s.Spec.Service.Name,
+							Namespace: s.Spec.Service.Namespace,
+						},
+					}
+					return e
+				}(),
+			},
+		},
```

Comment:
How about a case where the "status" key is present in the configmap but has a value other than "complete"?

Reply:
already covered in https://github.com/openshift/library-go/blob/master/pkg/operator/bootstrap/bootstrap_test.go#L20, do we want to duplicate that case here as well?

Reply:
I think it would be a good idea to either duplicate it or inject `IsBootstrapComplete`. We're testing the behavior of `preconditionsForEnabledAPIServices` here, so we can't rely on the knowledge that its implementation calls `IsBootstrapComplete` which has test coverage for this case.

Reply:
ok, extended the test.